google/cloud/storage/object_metadata.h

```diff
@@ -16,57 +16,161 @@
- * encryption keys, and other features.
- *
- * TODO(#693) - complete the implementation.
  */
-class ObjectMetadata : private internal::CommonMetadata {
+class ObjectMetadata : private internal::CommonMetadata<ObjectMetadata> {
  public:
-  ObjectMetadata() = default;
+  ObjectMetadata() : component_count_(0), generation_(0), size_(0) {}
+
+  static ObjectMetadata ParseFromJson(internal::nl::json const& json);
+  static ObjectMetadata ParseFromString(std::string const& payload);
 
-  static ObjectMetadata ParseFromJson(std::string const& payload);
+  // Please keep these in alphabetical order, that make it easier to verify we
+  // have actually implemented all of them.
+  std::vector<ObjectAccessControl> const& acl() const { return acl_; }
+  std::vector<ObjectAccessControl>& mutable_acl() { return acl_; }
+  ObjectMetadata& set_acl(std::vector<ObjectAccessControl> acl) {
+    acl_ = std::move(acl);
+    return *this;
+  }
+
+  std::string const& bucket() const { return bucket_; }
+
+  std::string const& cache_control() const { return cache_control_; }
+  ObjectMetadata& set_cache_control(std::string cache_control) {
+    cache_control_ = std::move(cache_control);
+    return *this;
+  }
+
+  long component_count() const { return component_count_; }
+
+  std::string content_disposition() const { return content_disposition_; }
+  ObjectMetadata& set_content_disposition(std::string value) {
+    content_disposition_ = std::move(value);
+    return *this;
+  }
+
+  std::string content_encoding() const { return content_encoding_; }
+  ObjectMetadata& set_content_encoding(std::string value) {
+    content_encoding_ = std::move(value);
+    return *this;
+  }
+
+  std::string content_language() const { return content_language_; }
+  ObjectMetadata& set_content_language(std::string value) {
+    content_language_ = std::move(value);
+    return *this;
+  }
+
+  std::string content_type() const { return content_type_; }
+  ObjectMetadata& set_content_type(std::string value) {
+    content_type_ = std::move(value);
+    return *this;
+  }
+
+  std::string const& crc32c() const { return crc32c_; }
+  CustomerEncryption const& customer_encryption() const {
+    return customer_encryption_;
+  }
+
+  using CommonMetadata::etag;
 
+  std::int64_t generation() const { return generation_; }
+
+  using CommonMetadata::id;
+  using CommonMetadata::kind;
+
+  std::string const& kms_key_name() const { return kms_key_name_; }
+  std::string const& md5_hash() const { return md5_hash_; }
+  std::string const& media_link() const { return media_link_; }
+
+  //@{
+  /// @name Accessors and modifiers to the `metadata` labels.
   std::size_t metadata_count() const { return metadata_.size(); }
   bool has_metadata(std::string const& key) const {
     return metadata_.end() != metadata_.find(key);
   }
   std::string const& metadata(std::string const& key) const {
     return metadata_.at(key);
   }
+  ObjectMetadata& delete_metadata(std::string const& key) {
```

Comment:
Probably worth adding a note that this succeeds (and does nothing) if the metadata item at the given key doesn't exist.

Reply:
Done.